### PR TITLE
Use hive.parquet.time-zone when reading INT64 timestamps in hive

### PR DESCRIPTION
**Pull request by mayankvadariya**
## Description

Parquet files having INT96 timestamps support `hive.parquet.time-zone` property for Hive connector. However, for UTC adjusted INT64 timestamps, setting `hive.parquet.time-zone` doesn't have any effect. This PR adds support for parquet reader to adjust INT64 timestamps to timezone defined by `hive.parquet.time-zone` when parquet logical annotation `isAdjustedToUTC` is set to true.

## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
# Hive
* Use `hive.parquet.time-zone` when reading UTC adjusted INT64 timestamps in parquet files. ({issue}`22577`)
```

### CODE REVIEW

**Comment by findinpath** [comment]
@mayankvadariya pls add a description containing the business value of the feature you're adding in this PR.

**Comment by findinpath** [comment]
nit: 

```
Error:  src/main/java/io/trino/plugin/deltalake/DeltaLakePageSourceProvider.java:[64,8] (imports) UnusedImports: Unused import - org.joda.time.DateTimeZone.
```

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSourceProvider.java [hunk 1]
```diff
@@ -249,7 +248,7 @@ public ConnectorPageSource createPageSource(
                 hiveColumnHandles.build(),
                 ImmutableList.of(parquetPredicate),
                 true,
-                parquetDateTimeZone,
+                UTC,
```
Comment (SemionPar) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSourceProvider.java:251]: *suggestion (non-blocking)*: this does not seem to be related to Hive - can/should this go in a separate commit? Also looks like we now are hardcoding some default - could you please add some context (in a new commit message)?

Comment (anusudarsan) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSourceProvider.java:250]: ya why this change?

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/parquet/ParquetTester.java [hunk 2]
```diff
@@ -816,7 +817,7 @@ private static void writeParquetColumnTrino(
                         .build(),
                 compressionCodec,
                 "test-version",
-                Optional.of(DateTimeZone.getDefault()),
+                Optional.of(UTC),
```
Comment (anusudarsan) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/parquet/ParquetTester.java:820]: are tests missing here for non-utc?

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/parquet/TestOnlyNulls.java [hunk 3]
```diff
@@ -56,13 +56,13 @@ public void testOnlyNulls()
         HiveColumnHandle column = createBaseColumn(columnName, 0, HiveType.toHiveType(columnType), columnType, REGULAR, Optional.empty());
 
         // match not null
-        try (ConnectorPageSource pageSource = createPageSource(SESSION, parquetFile, ImmutableList.of(column), TupleDomain.withColumnDomains(Map.of(column, notNull(columnType))))) {
+        try (ConnectorPageSource pageSource = createPageSourceWithUtcTz(SESSION, parquetFile, ImmutableList.of(column), TupleDomain.withColumnDomains(Map.of(column, notNull(columnType))))) {
```
Comment (anusudarsan) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/parquet/TestOnlyNulls.java:59]: same question on tests for non-utc

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/parquet/ParquetTester.java (hunk 2 above)
Comment (mayankvadariya) [reply plugin/trino-hive/src/test/java/io/trino/plugin/hive/parquet/ParquetTester.java:820]: Now that value of timezone takes effect test would fail as it tries to adjust the time with the given timezone.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSourceProvider.java [hunk 4]
```diff
@@ -115,7 +115,6 @@ public class DeltaLakePageSourceProvider
     private final FileFormatDataSourceStats fileFormatDataSourceStats;
     private final ParquetReaderOptions parquetReaderOptions;
     private final int domainCompactionThreshold;
-    private final DateTimeZone parquetDateTimeZone;
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSourceProvider.java:118]: Let's leave the removal of `delta.parquet.time-zone` for a follow-up PR.
Pls drop the current commit.

File: lib/trino-parquet/src/main/java/io/trino/parquet/reader/ColumnReaderFactory.java [hunk 5]
```diff
@@ -205,15 +205,15 @@ public ColumnReader create(PrimitiveField field, AggregatedMemoryContext aggrega
             }
             if (timestampType.isShort()) {
                 return switch (timestampAnnotation.getUnit()) {
-                    case MILLIS -> createColumnReader(field, valueDecoders::getInt64TimestampMillsToShortTimestampDecoder, LONG_ADAPTER, memoryContext);
-                    case MICROS -> createColumnReader(field, valueDecoders::getInt64TimestampMicrosToShortTimestampDecoder, LONG_ADAPTER, memoryContext);
-                    case NANOS -> createColumnReader(field, valueDecoders::getInt64TimestampNanosToShortTimestampDecoder, LONG_ADAPTER, memoryContext);
+                    case MILLIS -> createColumnReader(field, (encoding) -> valueDecoders.getInt64TimestampMillsToShortTimestampDecoder(encoding, timeZone), LONG_ADAPTER, memoryContext);
```
Comment (findinpath) [inline lib/trino-parquet/src/main/java/io/trino/parquet/reader/ColumnReaderFactory.java:208]:
```suggestion
case MILLIS -> createColumnReader(field, encoding -> valueDecoders.getInt64TimestampMillsToShortTimestampDecoder(encoding, timeZone), LONG_ADAPTER, memoryContext);
```

same for the other `(encoding) ->` occurences

File: lib/trino-parquet/src/main/java/io/trino/parquet/reader/decoders/ValueDecoders.java [hunk 6]
```diff
@@ -469,7 +469,7 @@ public void skip(int n)
         };
     }
 
-    public ValueDecoder<long[]> getInt64TimestampMillsToShortTimestampDecoder(ParquetEncoding encoding)
+    public ValueDecoder<long[]> getInt64TimestampMillsToShortTimestampDecoder(ParquetEncoding encoding, DateTimeZone timeZone)
```
Comment (findinpath) [inline lib/trino-parquet/src/main/java/io/trino/parquet/reader/decoders/ValueDecoders.java:472]: Pre-existing - separate commit 

Rename `getInt64TimestampMillsToShortTimestampDecoder` to `getInt64TimestampMillisToShortTimestampDecoder`

File: lib/trino-parquet/src/main/java/io/trino/parquet/reader/decoders/ValueDecoders.java [hunk 7]
```diff
@@ -534,14 +544,33 @@ public ValueDecoder<long[]> getInt64TimestampMicrosToShortTimestampDecoder(Parqu
...
         ValueDecoder<long[]> valueDecoder = getLongDecoder(encoding);
         if (precision == 6) {
-            return valueDecoder;
+            if (timeZone == DateTimeZone.UTC) {
+                return valueDecoder;
+            }
+            new InlineTransformDecoder<>(
+                    valueDecoder,
+                    (values, offset, length) -> {
+                        // decoded values are epochMicros, round to lower precision
```
Comment (findinpath) [inline lib/trino-parquet/src/main/java/io/trino/parquet/reader/decoders/ValueDecoders.java:553]: Comment doesn't make sense, the precision is already 6 - there's no lower precision.

File: lib/trino-parquet/src/main/java/io/trino/parquet/reader/decoders/ValueDecoders.java [hunk 8]
```diff
@@ -628,7 +662,12 @@ public void read(int[] values, int offset, int length)
                 delegate.read(buffer, 0, length);
                 // decoded values are epochMicros
                 for (int i = 0; i < length; i++) {
-                    encodeFixed12(buffer[i], 0, values, i + offset);
+                    long micros = buffer[i];
```
Comment (findinpath) [inline lib/trino-parquet/src/main/java/io/trino/parquet/reader/decoders/ValueDecoders.java:665]: don't you need to diferentiate between utc and non-utc?

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/parquet/ParquetUtil.java [hunk 9]
```diff
@@ -48,13 +48,25 @@ final class ParquetUtil
 {
     private ParquetUtil() {}
 
-    public static ConnectorPageSource createPageSource(ConnectorSession session, File parquetFile, List<String> columnNames, List<Type> columnTypes)
+    public static ConnectorPageSource createPageSourceWithUtcTz(ConnectorSession session, File parquetFile, List<String> columnNames, List<Type> columnTypes)
```
Comment (findinpath) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/parquet/ParquetUtil.java:51]: Avoid shortcuts in naming the methods

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/parquet/TestOnlyNulls.java [hunk 10]
```diff
@@ -56,13 +56,13 @@ public void testOnlyNulls()
         HiveColumnHandle column = createBaseColumn(columnName, 0, HiveType.toHiveType(columnType), columnType, REGULAR, Optional.empty());
 
         // match not null
-        try (ConnectorPageSource pageSource = createPageSource(SESSION, parquetFile, ImmutableList.of(column), TupleDomain.withColumnDomains(Map.of(column, notNull(columnType))))) {
```
Comment (findinpath) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/parquet/TestOnlyNulls.java:59]: This test would still pass with the old `io.trino.plugin.hive.parquet.ParquetUtil#createPageSource(io.trino.spi.connector.ConnectorSession, java.io.File, java.util.List<io.trino.plugin.hive.HiveColumnHandle>, io.trino.spi.predicate.TupleDomain<io.trino.plugin.hive.HiveColumnHandle>)` method.

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/parquet/TestTimestamp.java [hunk 11]
```diff
@@ -71,29 +79,36 @@ public void testTimestampBackedByInt64()
             MessageType parquetSchema = parseMessageType("message hive_timestamp { optional int64 test (" + logicalAnnotation + "); }");
 
             Iterable<Long> writeNullableDictionaryValues = limit(cycle(asList(1L, null, 3L, 5L, null, null, null, 7L, 11L, null, 13L, 17L)), 30_000);
-            testRoundTrip(parquetSchema, writeNullableDictionaryValues, timestamp);
+            testRoundTrip(parquetSchema, writeNullableDictionaryValues, timestamp, dateTimeZone);
 
             Iterable<Long> writeDictionaryValues = limit(cycle(asList(1L, 3L, 5L, 7L, 11L, 13L, 17L)), 30_000);
```
Comment (findinpath) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/parquet/TestTimestamp.java:84]: Why aren't we using also values which are closer the current times (not only very close to 1970-01-01) for testing?

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/parquet/ParquetTester.java [hunk 12]
```diff
@@ -478,7 +479,7 @@ private void assertFileContents(
             List<Type> columnTypes)
             throws IOException
     {
-        try (ConnectorPageSource pageSource = createPageSource(
+        try (ConnectorPageSource pageSource = createPageSourceWithUtcTz(
```
Comment (raunaqmorarka) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/parquet/ParquetTester.java:482]: Rather than changing this for all tests, we should add a boolean flag `readWithUtcTimezone` to `assertRoundTrip` and have specific tests enabling that. This way we can test with UTC as well as default and see where exactly the override was needed.

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/parquet/TestTimestamp.java (hunk 11 above)
Comment (mayankvadariya) [reply plugin/trino-hive/src/test/java/io/trino/plugin/hive/parquet/TestTimestamp.java:85]: `AbstractTestParquetReader#testTimestamp` simulates round trip with current timestamp.

File: lib/trino-parquet/src/main/java/io/trino/parquet/reader/decoders/ValueDecoders.java [hunk 13]
```diff
@@ -575,12 +604,17 @@ public ValueDecoder<long[]> getInt64TimestampNanosToShortTimestampDecoder(Parque
                 (values, offset, length) -> {
                     // decoded values are epochNanos, round to lower precision and convert to epochMicros
                     for (int i = offset; i < offset + length; i++) {
-                        values[i] = round(values[i], 9 - precision) / NANOSECONDS_PER_MICROSECOND;
+                        if (timeZone == DateTimeZone.UTC) {
+                            values[i] = round(values[i], 9 - precision) / NANOSECONDS_PER_MICROSECOND;
+                        }
+                        else {
+                            values[i] = timeZone.convertUTCToLocal(round(values[i], 9 - precision)) / NANOSECONDS_PER_MICROSECOND;
```
[the remark this answers is not in the document]
Comment (mayankvadariya) [reply lib/trino-parquet/src/main/java/io/trino/parquet/reader/decoders/ValueDecoders.java]: Ya that's correct, made required changes.

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/parquet/ParquetTester.java (hunk 12 above)
Comment (mayankvadariya) [reply plugin/trino-hive/src/test/java/io/trino/plugin/hive/parquet/ParquetTester.java]: Now ParquetTester doesn't require any change after changing the parquet writer.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSourceProvider.java (hunk 1 above)
Comment (mayankvadariya) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSourceProvider.java:250]: Explanation https://github.com/trinodb/trino/pull/22577#discussion_r1680404943

File: lib/trino-parquet/src/main/java/io/trino/parquet/writer/valuewriter/TimestampMillisValueWriter.java [hunk 14]
```diff
@@ -14,31 +14,52 @@
...
+    public TimestampMillisValueWriter(ValuesWriter valuesWriter, Type type, PrimitiveType parquetType, DateTimeZone parquetTimeZone)
     {
         super(parquetType, valuesWriter);
-        this.type = requireNonNull(type, "type is null");
+        requireNonNull(type, "type is null");
+        checkArgument(
+                type instanceof TimestampType && ((TimestampType) type).getPrecision() <= 9,
+                "type %s is not a TimestampType with precision <= 9",
+                type);
+        this.timestampType = (TimestampType) type;
```
Comment (raunaqmorarka) [inline lib/trino-parquet/src/main/java/io/trino/parquet/writer/valuewriter/TimestampMillisValueWriter.java:43]: Why is this change needed ?

File: lib/trino-parquet/src/main/java/io/trino/parquet/writer/valuewriter/TimestampNanosValueWriter.java [hunk 15]
```diff
@@ -16,35 +16,63 @@
...
+    public TimestampNanosValueWriter(ValuesWriter valuesWriter, Type type, PrimitiveType parquetType, DateTimeZone parquetTimeZone)
     {
         super(parquetType, valuesWriter);
-        this.type = requireNonNull(type, "type is null");
+        requireNonNull(type, "type is null");
+        checkArgument(
+                type instanceof TimestampType && ((TimestampType) type).getPrecision() <= 9,
+                "type %s is not a TimestampType with precision <= 9",
+                type);
+        this.timestampType = (TimestampType) type;
```
Comment (raunaqmorarka) [inline lib/trino-parquet/src/main/java/io/trino/parquet/writer/valuewriter/TimestampNanosValueWriter.java:49]: Is this change needed ?

File: lib/trino-parquet/src/main/java/io/trino/parquet/writer/valuewriter/TimestampNanosValueWriter.java [hunk 16]
```diff
@@ -16,35 +16,63 @@
...
     {
         for (int i = 0; i < block.getPositionCount(); i++) {
             if (!block.isNull(i)) {
-                LongTimestamp value = (LongTimestamp) type.getObject(block, i);
-                long epochNanos = multiplyExact(value.getEpochMicros(), NANOSECONDS_PER_MICROSECOND) +
-                        LongMath.divide(value.getPicosOfMicro(), PICOSECONDS_PER_NANOSECOND, UNNECESSARY);
+                LongTimestamp value = (LongTimestamp) timestampType.getObject(block, i);
+                long epochNanos;
+                if(parquetTimeZone == DateTimeZone.UTC) {
+                    epochNanos = multiplyExact(value.getEpochMicros(), NANOSECONDS_PER_MICROSECOND) +
```
Comment (raunaqmorarka) [inline lib/trino-parquet/src/main/java/io/trino/parquet/writer/valuewriter/TimestampNanosValueWriter.java:65]: Similar comment as in TimestampMillisValueWriter

File: lib/trino-parquet/src/main/java/io/trino/parquet/writer/ParquetWriters.java [hunk 17]
```diff
@@ -103,7 +103,7 @@ static PrimitiveValueWriter getValueWriter(ValuesWriter valuesWriter, Type type,
             return new IntegerValueWriter(valuesWriter, type, parquetType);
         }
```
Comment (raunaqmorarka) [inline lib/trino-parquet/src/main/java/io/trino/parquet/writer/ParquetWriters.java:104]: We also need to ensure that `parquetTimeZone` is hard-coded to UTC when parquet writer is used through hudi/iceberg/delta, just like we're doing for the reader.

Comment (mayankvadariya) [reply lib/trino-parquet/src/main/java/io/trino/parquet/writer/ParquetWriters.java]: Good point.

https://github.com/trinodb/trino/actions/runs/10014094980/job/27683136191?pr=22577#step:5:2434

Aluxio caching test is failing with defaulting parquet writer to UTC in delta
https://github.com/trinodb/trino/pull/22577/commits/306315e5b10b3e60630c665ff64abe925077457b#diff-836fd9091213321cf465d1d29d45862bf927cc60daaa7b4e28713080010b6d4dR500-R501

File: lib/trino-parquet/src/main/java/io/trino/parquet/reader/decoders/ValueDecoders.java [hunk 18]
```diff
@@ -694,11 +745,22 @@ public void read(int[] values, int offset, int length)
...
                     long epochNanos = buffer[i];
-                    encodeFixed12(
-                            floorDiv(epochNanos, NANOSECONDS_PER_MICROSECOND),
-                            floorMod(epochNanos, NANOSECONDS_PER_MICROSECOND) * PICOSECONDS_PER_NANOSECOND,
-                            values,
-                            i + offset);
+                    if (timeZone == DateTimeZone.UTC) {
+                        encodeFixed12(
+                                floorDiv(epochNanos, NANOSECONDS_PER_MICROSECOND),
+                                floorMod(epochNanos, NANOSECONDS_PER_MICROSECOND) * PICOSECONDS_PER_NANOSECOND,
```
Comment (raunaqmorarka) [inline lib/trino-parquet/src/main/java/io/trino/parquet/reader/decoders/ValueDecoders.java:751]: Extract common variable

File: lib/trino-parquet/src/main/java/io/trino/parquet/writer/ParquetWriters.java [hunk 19]
```diff
@@ -125,21 +126,21 @@ static PrimitiveValueWriter getValueWriter(ValuesWriter valuesWriter, Type type,
...
         if (type instanceof TimestampType) {
+            checkArgument(parquetTimeZone.isPresent(), "parquetTimeZone must be provided for timestamps");
             if (parquetType.getPrimitiveTypeName().equals(PrimitiveType.PrimitiveTypeName.INT96)) {
-                checkArgument(parquetTimeZone.isPresent(), "parquetTimeZone must be provided for INT96 timestamps");
                 return new Int96TimestampValueWriter(valuesWriter, type, parquetType, parquetTimeZone.get());
             }
             if (TIMESTAMP_MILLIS.equals(type)) {
                 verifyParquetType(type, parquetType, TimestampLogicalTypeAnnotation.class, isTimestamp(LogicalTypeAnnotation.TimeUnit.MILLIS));
-                return new TimestampMillisValueWriter(valuesWriter, type, parquetType);
+                return new TimestampMillisValueWriter(valuesWriter, type, parquetType, parquetTimeZone.get());
```
Comment (raunaqmorarka) [inline lib/trino-parquet/src/main/java/io/trino/parquet/writer/ParquetWriters.java:135]: We need to update `io.trino.parquet.writer.ParquetSchemaConverter#getPrimitiveType` to set isAdjustedToUTC to true in `LogicalTypeAnnotation.timestampType` if non-UTC timezone is configured.  ParquetSchemaConverter should always receive UTC from non-hive connectors.

File: lib/trino-parquet/src/main/java/io/trino/parquet/writer/valuewriter/TimestampNanosValueWriter.java [hunk 20]
```diff
@@ -45,6 +51,14 @@ public void write(Block block)
                 LongTimestamp value = (LongTimestamp) type.getObject(block, i);
                 long epochNanos = multiplyExact(value.getEpochMicros(), NANOSECONDS_PER_MICROSECOND) +
                         LongMath.divide(value.getPicosOfMicro(), PICOSECONDS_PER_NANOSECOND, UNNECESSARY);
+                if (parquetTimeZone != DateTimeZone.UTC) {
+                    long nanos = multiplyExact(value.getEpochMicros(), NANOSECONDS_PER_MICROSECOND) +
+                            LongMath.divide(value.getPicosOfMicro(), PICOSECONDS_PER_NANOSECOND, UNNECESSARY);
```
Comment (raunaqmorarka) [inline lib/trino-parquet/src/main/java/io/trino/parquet/writer/valuewriter/TimestampNanosValueWriter.java:56]: Looks like unnecessary repeat of `epochNanos`

File: lib/trino-parquet/src/main/java/io/trino/parquet/writer/valuewriter/TimestampMicrosValueWriter.java [hunk 21]
```diff
@@ -0,0 +1,60 @@
...
+    public void write(Block block)
+    {
+        for (int i = 0; i < block.getPositionCount(); i++) {
+            if (!block.isNull(i)) {
+                long value;
+                if (parquetTimeZone == UTC) {
+                    value = type.getLong(block, i);
+                }
+                else {
+                    long micros = type.getLong(block, i);
```
Comment (raunaqmorarka) [inline lib/trino-parquet/src/main/java/io/trino/parquet/writer/valuewriter/TimestampMicrosValueWriter.java:51]:
```
long epochMicros = type.getLong(block, i);
if (parquetTimeZone != UTC) {
     epochMicros = ...
}
```

File: lib/trino-parquet/src/main/java/io/trino/parquet/writer/valuewriter/TimestampMillisValueWriter.java (hunk 14 above)
Comment (mayankvadariya) [reply lib/trino-parquet/src/main/java/io/trino/parquet/writer/valuewriter/TimestampMillisValueWriter.java]: Had adapted it from INT96 writer. Removed.

File: lib/trino-parquet/src/main/java/io/trino/parquet/reader/ColumnReaderFactory.java [hunk 22]
```diff
@@ -205,15 +205,15 @@ public ColumnReader create(PrimitiveField field, AggregatedMemoryContext aggrega
             }
             if (timestampType.isShort()) {
```
[the remark this answers is not in the document]
Comment (mayankvadariya) [reply lib/trino-parquet/src/main/java/io/trino/parquet/reader/ColumnReaderFactory.java:207]: Done

File: lib/trino-parquet/src/main/java/io/trino/parquet/writer/ParquetWriters.java [hunk 23]
```diff
@@ -126,20 +127,20 @@ static PrimitiveValueWriter getValueWriter(ValuesWriter valuesWriter, Type type,
...
         if (type instanceof TimestampType) {
             if (parquetType.getPrimitiveTypeName().equals(PrimitiveType.PrimitiveTypeName.INT96)) {
-                checkArgument(parquetTimeZone.isPresent(), "parquetTimeZone must be provided for INT96 timestamps");
+                checkArgument(parquetTimeZone.isPresent(), "parquetTimeZone must be provided for timestamps");
                 return new Int96TimestampValueWriter(valuesWriter, type, parquetType, parquetTimeZone.get());
             }
             if (TIMESTAMP_MILLIS.equals(type)) {
                 verifyParquetType(type, parquetType, TimestampLogicalTypeAnnotation.class, isTimestamp(LogicalTypeAnnotation.TimeUnit.MILLIS));
-                return new TimestampMillisValueWriter(valuesWriter, type, parquetType);
+                return new TimestampMillisValueWriter(valuesWriter, type, parquetType, parquetTimeZone);
```
Comment (raunaqmorarka) [inline lib/trino-parquet/src/main/java/io/trino/parquet/writer/ParquetWriters.java:135]: I've had a re-think about the writer changes. Now that we're deciding on timezone conversion based on parquet annotation rather than just the time zone config, we don't have the problem of data not round tripping properly in tests.
I had also found that in Hive 4, when int64 timestamps (`set hive.parquet.write.int64.timestamp=true;`) are written, they're not adjusted to UTC.
Let's drop the writer changes for now.

File: lib/trino-parquet/src/main/java/io/trino/parquet/reader/decoders/ValueDecoders.java [hunk 24]
```diff
@@ -628,7 +669,17 @@ public void read(int[] values, int offset, int length)
                 delegate.read(buffer, 0, length);
                 // decoded values are epochMicros
                 for (int i = 0; i < length; i++) {
-                    encodeFixed12(buffer[i], 0, values, i + offset);
+                    long epochMicros = buffer[i];
+                    if (timeZone == DateTimeZone.UTC) {
+                        encodeFixed12(buffer[i], 0, values, i + offset);
+                    }
+                    else {
```
[the remark this answers is not in the document]
Comment (mayankvadariya) [reply lib/trino-parquet/src/main/java/io/trino/parquet/reader/decoders/ValueDecoders.java:701]: Using `localMillis` here and everywhere.

File: lib/trino-parquet/src/main/java/io/trino/parquet/reader/decoders/ValueDecoders.java [hunk 25]
```diff
@@ -575,12 +604,19 @@ public ValueDecoder<long[]> getInt64TimestampNanosToShortTimestampDecoder(Parque
...
                     // decoded values are epochNanos, round to lower precision and convert to epochMicros
                     for (int i = offset; i < offset + length; i++) {
-                        values[i] = round(values[i], 9 - precision) / NANOSECONDS_PER_MICROSECOND;
+                        long epochNanos = round(values[i], 9 - precision);
+                        if (timeZone == DateTimeZone.UTC) {
+                            values[i] = epochNanos / NANOSECONDS_PER_MICROSECOND;
+                        }
+                        else {
+                            values[i] = (timeZone.convertUTCToLocal(floorDiv(epochNanos, NANOSECONDS_PER_MICROSECOND * MICROSECONDS_PER_MILLISECOND)) / NANOSECONDS_PER_MICROSECOND)
+                                    + floorDiv(epochNanos, NANOSECONDS_PER_MICROSECOND);
```
[the remark this answers is not in the document]
Comment (mayankvadariya) [reply lib/trino-parquet/src/main/java/io/trino/parquet/reader/decoders/ValueDecoders.java]: yes catch

File: lib/trino-parquet/src/main/java/io/trino/parquet/reader/decoders/ValueDecoders.java [hunk 26]
```diff
@@ -575,12 +604,19 @@ public ValueDecoder<long[]> getInt64TimestampNanosToShortTimestampDecoder(Parque
                 (values, offset, length) -> {
                     // decoded values are epochNanos, round to lower precision and convert to epochMicros
                     for (int i = offset; i < offset + length; i++) {
-                        values[i] = round(values[i], 9 - precision) / NANOSECONDS_PER_MICROSECOND;
+                        long epochNanos = round(values[i], 9 - precision);
+                        if (timeZone == DateTimeZone.UTC) {
+                            values[i] = epochNanos / NANOSECONDS_PER_MICROSECOND;
+                        }
+                        else {
+                            values[i] = (timeZone.convertUTCToLocal(floorDiv(epochNanos, NANOSECONDS_PER_MICROSECOND * MICROSECONDS_PER_MILLISECOND)) / NANOSECONDS_PER_MICROSECOND)
```
[the remark this answers is not in the document]
Comment (mayankvadariya) [reply lib/trino-parquet/src/main/java/io/trino/parquet/reader/decoders/ValueDecoders.java]: good call

**Comment by mayankvadariya** [comment]
Resolved conflicts and addressed comments.

File: lib/trino-parquet/src/main/java/io/trino/parquet/writer/ParquetWriters.java (hunk 23 above)
Comment (mayankvadariya) [reply lib/trino-parquet/src/main/java/io/trino/parquet/writer/ParquetWriters.java:135]: I think without https://github.com/trinodb/trino/pull/22577/files#diff-3b64954c48d2c0059a9156e9ef84a6b66d58c436ea593753a78cd27a604c92feR128-R328 changes, writer will write time without converting to to UTC and reader will consider that time in UTC and then, convert to local, do we need to make any other changes?

Comment (raunaqmorarka) [reply lib/trino-parquet/src/main/java/io/trino/parquet/writer/ParquetWriters.java:135]: Let's say we start by reading adjustedToUTC INT64 timestamp, we will convert from UTC to local. If we write it out as iceberg or delta, then it's INT64 timestamps in local time, which is expected for iceberg/delta tables. It we write it out as hive table, then it will be int96 timestamp which is always converted to UTC.
So I don't think there is a problem with skipping the writer changes.

**Review by raunaqmorarka — COMMENTED** [review]
please squash the fixups

File: lib/trino-parquet/src/main/java/io/trino/parquet/writer/ParquetWriters.java (hunk 23 above)
Comment (mayankvadariya) [reply lib/trino-parquet/src/main/java/io/trino/parquet/writer/ParquetWriters.java:135]: Thanks for clarifying the round trip for table formats. However, Parquet round trip test `AbstractTestParquetReader#testTimestamp` is failing without the mentioned change.

Comment (raunaqmorarka) [reply lib/trino-parquet/src/main/java/io/trino/parquet/writer/ParquetWriters.java]: i pushed a fixup commit which removes the writer changes, AbstractTestParquetReader#testTimestamp is passing

File: lib/trino-parquet/src/main/java/io/trino/parquet/writer/valuewriter/TimestampMicrosValueWriter.java [hunk 27]
```diff
@@ -0,0 +1,58 @@
...
+
+import java.util.Optional;
+
+import static io.trino.spi.type.Timestamps.MICROSECONDS_PER_MILLISECOND;
+import static java.lang.Math.floorDiv;
+import static java.lang.Math.floorMod;
+import static java.util.Objects.requireNonNull;
+import static org.joda.time.DateTimeZone.UTC;
+
+public class TimestampMicrosValueWriter
```
Comment (raunaqmorarka) [inline lib/trino-parquet/src/main/java/io/trino/parquet/writer/valuewriter/TimestampMicrosValueWriter.java:30]: This class should be removed now

Comment (mayankvadariya) [reply lib/trino-parquet/src/main/java/io/trino/parquet/writer/valuewriter/TimestampMicrosValueWriter.java]: yes, I had it removed in some fixup. Removed

File: lib/trino-parquet/src/main/java/io/trino/parquet/writer/ParquetWriters.java (hunk 23 above)
Comment (mayankvadariya) [reply lib/trino-parquet/src/main/java/io/trino/parquet/writer/ParquetWriters.java]: I didn't revert the `isAdjustedToUTC` changes in `ParquetSchemaConverter`

**Review by raunaqmorarka — APPROVED** [review]
